pubsub/natspubsub/nats.go

```diff
@@ -0,0 +1,263 @@
+// Copyright 2019 The Go Cloud Development Kit Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package natspubsub provides a pubsub implementation for NATS.io.
+// Use OpenTopic to construct a *pubsub.Topic, and/or OpenSubscription
+// to construct a *pubsub.Subscription.
+//
+// As
+//
+// natspubsub exposes the following types for As:
+//  - Topic: *nats.Conn
+//  - Subscription: *nats.Subscription
+//  - Message: *nats.Msg
+
+package natspubsub // import "gocloud.dev/pubsub/natspubsub"
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/nats-io/go-nats"
+	"github.com/ugorji/go/codec"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+var errNotExist = errors.New("natspubsub: topic does not exist")
+
+type topic struct {
+	nc   *nats.Conn
+	subj string
+}
+
+// For encoding we use msgpack from github.com/ugorji/go.
+// it was already imported in go-cloud and is reasonably performant.
+var mh codec.MsgpackHandle
+
+func init() {
+	// driver.Message.Metadata type
+	dm := driver.Message{}
+	mh.MapType = reflect.TypeOf(dm.Metadata)
+	mh.ExplicitRelease = true
+}
+
+// We define our own version of message here for encoding that
+// only encodes Body and Metadata. Otherwise we would have to
+// add codec decorations to driver.Message.
+type encMsg struct {
+	Body     []byte            `codec:",omitempty"`
+	Metadata map[string]string `codec:",omitempty"`
+}
+
+// OpenTopic returns a *pubsub.Topic for use with NATS.
+// We delay checking for the proper syntax here.
+func OpenTopic(nc *nats.Conn, topicName string) *pubsub.Topic {
+	return pubsub.NewTopic(openTopic(nc, topicName))
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(nc *nats.Conn, topicName string) driver.Topic {
+	return &topic{nc, topicName}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
+	if t == nil || t.nc == nil {
+		return errNotExist
```

Comment:
Wouldn't it make more sense to say it's uninitialized, rather than that it doesn't exist?

Reply:
Possibly, copied what others were doing here. Will fix.